classes/mail/Mailable.inc.php

```diff
@@ -0,0 +1,272 @@
+     */
+    public static function getGroupIds() : array
+    {
+        return static::$groupIds;
+    }
+
+    /**
+     * Method's implementation is required for Mailable to be sent according to Laravel docs
+     * @see \Illuminate\Mail\Mailable::send(), https://laravel.com/docs/7.x/mail#writing-mailables
+     */
+    public function build() : self
+    {
+        return $this;
+    }
+
+    /**
+     * Allow data to be passed to the subject
+     * @param \Illuminate\Mail\Message $message
+     * @throws Exception
+     */
+    protected function buildSubject($message) : self
+    {
+        if (!$this->subject) {
+            throw new Exception('Subject isn\'t specified in ' . static::class);
+        }
+
+        $subject = app('mailer')->compileParams($this->subject, $this->viewData);
+        $message->subject($subject);
+
+        return $this;
+    }
+
+    /**
+     * Returns variables map associated with a specific object,
+     * variables names should be unique
+     * @return string[]
+     */
+    protected static function templateVariablesMap() : array
+    {
+        return
+            [
+                Site::class => SiteEmailVariable::class,
+                Context::class => ContextEmailVariable::class,
+                PKPSubmission::class => SubmissionEmailVariable::class,
+                ReviewAssignment::class => ReviewAssignmentEmailVariable::class,
+                StageAssignment::class => StageAssignmentEmailVariable::class,
+                QueuedPayment::class => QueuedPaymentEmailVariable::class,
+            ];
+    }
+
+    /**
+     * Scans arguments to retrieve variables which can be assigned to the template of the email
+     */
+    protected function setData(array $args) : void
+    {
+        $map = static::templateVariablesMap();
+
+        foreach ($args as $arg) {
+            foreach ($map as $className => $assoc) {
+                if (is_a($arg, $className)) {
+                    $assocVariable = new $assoc($arg);
+                    $this->viewData = array_merge(
+                        $this->viewData,
+                        $assocVariable->getValue()
+                    );
+                    continue 2;
+                }
+            }
+
+            // Give up, object isn't mapped
+            $type = is_object($arg) ? get_class($arg) : gettype($arg);
+            throw new InvalidArgumentException($type . ' argument passed to the ' . static::class . ' constructor isn\'t associated with template variables');
+        }
+    }
+
+    /**
+     * Retrieves array of variables that can be assigned to email templates
+     * @return array ['variableName' => description]
+     */
+    public static function getVariables() : array
+    {
+        $args = static::getParamsClass(static::getConstructor());
+        $map = static::templateVariablesMap();
+        $variables = [];
+
+        $traits = class_uses(static::class);
```

Comment:
Neat! I didn't know about this method. It looks like it doesn't get traits on parent classes, which is probably going to bite us down the line. There are a couple of example recursive functions you could use in the comments on the docs here: https://www.php.net/manual/en/function.class-uses.php

Reply:
Thanks, this should do the trick: https://github.com/pkp/pkp-lib/pull/7221/commits/45923c3ae7111a3438f6d756a62a7c8d86ca995a

Reply:
I'm glad to see this is using `::class` rather than depending overmuch on string-based class names (and I wish there was a closer match to `instanceof` for traits that didn't use string names at all); working with string-based classnames is prone to regression (e.g. when we introduced namespaces) and I tried to transform any cases of e.g. `get_class` with `instanceof`. I really like the use of traits for this purpose.